Rules/AvoidOverwritingBuiltInCmdlets.cs

```diff
@@ -0,0 +1,264 @@
+using System.Text.RegularExpressions;
+using Microsoft.Windows.PowerShell.ScriptAnalyzer.Generic;
+
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Windows.PowerShell.ScriptAnalyzer.BuiltinRules
+{
+    /// <summary>
+    /// AvoidOverwritingBuiltInCmdlets: Checks if a script overwrites a cmdlet that comes with PowerShell
+    /// </summary>
+#if !CORECLR
+    [Export(typeof(IScriptRule))]
+#endif
+    /// <summary>
+    /// A class to check if a script overwrites a cmdlet that comes with PowerShell
+    /// </summary>
+    public class AvoidOverwritingBuiltInCmdlets : ConfigurableRule
+    {
+        /// <summary>
+        /// Specify the version of PowerShell to compare against since different versions of PowerShell
+        /// ship with different sets of built in cmdlets. The default value for PowerShellVersion is
+        /// "core-6.1.0-windows" if PowerShell 6 or later is installed, and "desktop-5.1.14393.206-windows"
+        /// if it is not. The version specified aligns with a JSON file in `/path/to/PSScriptAnalyzerModule/Settings`.
+        /// These files are of the form, `PSEDITION-PSVERSION-OS.json` where `PSEDITION` can be either `Core` or
+        /// `Desktop`, `OS` can be either `Windows`, `Linux` or `MacOS`, and `Version` is the PowerShell version.
+        /// </summary>
+        [ConfigurableRuleProperty(defaultValue: "")]
+        public string[] PowerShellVersion { get; set; }
+        private readonly Dictionary<string, HashSet<string>> _cmdletMap;
+
+
+        /// <summary>
+        /// Construct an object of AvoidOverwritingBuiltInCmdlets type.
+        /// </summary>
+        public AvoidOverwritingBuiltInCmdlets()
+        {
+            _cmdletMap = new Dictionary<string, HashSet<string>>();
+            Enable = true;  // Enable rule by default
+        }
+
+
+        /// <summary>
+        /// Analyzes the given ast to find the [violation]
+        /// </summary>
+        /// <param name="ast">AST to be analyzed. This should be non-null</param>
+        /// <param name="fileName">Name of file that corresponds to the input AST.</param>
+        /// <returns>A an enumerable type containing the violations</returns>
+        public override IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
+        {
+            if (ast == null)
+            {
+                throw new ArgumentNullException(nameof(ast));
+            }
+
+            var diagnosticRecords = new List<DiagnosticRecord>();
+
+            IEnumerable<FunctionDefinitionAst> functionDefinitions = ast.FindAll(testAst => testAst is FunctionDefinitionAst, true).OfType<FunctionDefinitionAst>();
+            if (functionDefinitions.Count() < 1)
+            {
+                // There are no function definitions in this AST and so it's not worth checking the rest of this rule
+                return diagnosticRecords;
+            }
+
+
+            string versionTest = string.Join("", PowerShellVersion);
```

Comment:
I think you can move this into the if statement and in the if statement just assert against the `Count`/`Length` of `PowerShellVersion` as this would be a bit more efficient and easier to read.

Reply:
If it's not used in the if statement, the `versionTest` variable doesn't have a purpose, so I've removed it. My trouble is that if you don't pass a value for `PowerShellVersion`, it ends up as a one item array whose only value is an empty string. So, it doesn't have a count of 0. See what you think of my change. I'm not sure I prefer it, but maybe it's better.